service/src/main/java/skills/services/quiz/QuizDefService.groovy

```diff
@@ -167,6 +170,81 @@ class QuizDefService {
         )
     }
 
+    @Transactional()
+    QuizDefResult copyQuiz(String originalQuizId, String newQuizId, QuizDefRequest quizDefRequest, String userIdParam = null) {
+        validateQuizDefRequest(newQuizId, quizDefRequest)
+
+        newQuizId = InputSanitizer.sanitize(newQuizId)
+        quizDefRequest.name = InputSanitizer.sanitize(quizDefRequest.name)?.trim()
+        quizDefRequest.description = StringUtils.trimToNull(InputSanitizer.sanitize(quizDefRequest.description))
+
+        lockingService.lockQuizDefs()
+
+        String userId = userIdParam ?: userInfoService.getCurrentUserId()
+        QuizDefWithDescription quizDefWithDescription = copyQuizDef(originalQuizId, newQuizId, quizDefRequest, userId)
+        copyQuestions(originalQuizId, newQuizId)
+        quizSettingsService.copySettings(originalQuizId, newQuizId)
+
+        attachmentService.updateAttachmentsFoundInMarkdown(quizDefRequest.description, null, newQuizId, null)
+
+        accessSettingsStorageService.addQuizDefUserRole(userId, newQuizId, RoleName.ROLE_QUIZ_ADMIN)
+
+        log.debug("Copied [{}]", quizDefWithDescription)
+
+        userActionsHistoryService.saveUserAction(new UserActionInfo(
+                action: DashboardAction.Create,
+                item: DashboardItem.Quiz,
+                actionAttributes: quizDefWithDescription,
+                itemId: quizDefWithDescription.quizId,
+                itemRefId: quizDefWithDescription.id,
+                quizId: quizDefWithDescription.quizId,
+        ))
+
+        QuizDef updatedDef = quizDefRepo.findByQuizIdIgnoreCase(quizDefWithDescription.quizId)
+        return convert(updatedDef)
+    }
+
+    QuizDefWithDescription copyQuizDef(String originalQuizId, String newQuizId, QuizDefRequest quizDefRequest, String userId) {
+        QuizDefWithDescription quizDefWithDescription = originalQuizId ? quizDefWithDescRepo.findByQuizIdIgnoreCase(originalQuizId) : null
+        if (!quizDefWithDescription || !quizDefWithDescription.quizId.equalsIgnoreCase(originalQuizId)) {
+            serviceValidatorHelper.validateQuizIdDoesNotExist(newQuizId)
+        }
+        if (!quizDefWithDescription || !quizDefWithDescription.name.equalsIgnoreCase(quizDefWithDescription.name)) {
+            serviceValidatorHelper.validateQuizNameDoesNotExist(quizDefRequest.name, newQuizId)
+        }
+
+        quizDefWithDescription = new QuizDefWithDescription(quizId: newQuizId, name: quizDefRequest.name,
+                description: quizDefRequest.description, type: QuizDefParent.QuizType.valueOf(quizDefRequest.type))
+        log.debug("Created quiz [{}]", quizDefWithDescription)
+
+        if (!userInfoService.isCurrentUserASuperDuperUser()) {
+            createdResourceLimitsValidator.validateNumQuizDefsCreated(userId)
```

Comment:
suggesting to remove this check; if the threshold is decreased after project is created I am thinking we still allow the copy to succeed; please create a service tests for this case as well